README.md
# kochava_postback

Simulation of a method Kochava uses to distribute data to third parties in real time.

## Planning

The first step was to set up my development environment. Based upon the criteria, I initially created one monorepo with a folder for the delivery and ingest agents. Normally, I would separate these into two different repositories (like how I did with the GU AR Walking Tour Project). The monorepo solution ended up working well for this project. Inside the ingestion and delivery folders are Dockerfile's specific to that agent.

Central to the entire project is the Redis queue. This is managed using Redis Lists. The operations LPush and BRPop create a FIFO queue going from left to right. BRPop ensures that access is blocked to the list until the operation is complete. This helps prevent any access timing issues for the specific key being popped from. The ingest program (PHP) LPush'es the postback object when it is recieved from a POST request. As fast as possible, the delivery agent (Go) pulls objects from out of the queue and quickly processes them, finally sending a request to the formatted response URL.

## Problems Encountered

The first major issue that required me to make a design choice was the Unmatched Key specification (in Extra Credit). In a language like Python, JSON objects can be natively converted to Dictionaries, whereas Go requires a stricter Type-safe implementation using structs. The only other way in Go is to map the values, which increases complexity and removes the type-safety of the application. Given that, and further information from [Maps vs. Structs for JSON (Ben Cane)](https://bencane.com/2020/12/08/maps-vs-structs-for-json/), I chose to implement a struct that represents the ingested JSON. If more fields are needed in the future (beyond "location" and "mascot") then they should be added to the struct. The value for unmatched keys are now set to the default as specified in the constructURL() utility. This ensures that other values can be safely passed through the ingest and out through delivery while maintaining safe and compliant URL's.

Another problem was encountered in setting up PHP's Docker container. I have a very small amount of familiarity with PHP and the installation on my computer took a fair amount of effort and troubleshooting. I wanted to streamline the install in Docker though, and eventually settled on a PHP image that includes the Apache webserver. This made the install much smoother, and only required installing the Redis extension on top of the base image. I have found that this works well, especially now that I am utilizing a multi-container architecture (See Docker section below).

## Docker

At the beginning I was slightly confused how I would be able to structure and distribute this project from a single Docker container, like Ubuntu:14.04. The Docker documentation says, "It is generally recommended that you separate areas of concern by using one service per container". This is the method that I prefer, as it allows for better horizontal scaling and much safer failures (that don't take the entire application down). In the end, I settled for using three separate containers, one for delivery, one for ingestion, and the other for Redis. These three containers can be easily managed by the scripts in the "scripts" folder as well as through the Makefile.

## Prior to Testing

Ensure that the addresses and ports in the files described in [DOCUMENTATION.md](DOCUMENTATION.md) have been changed to correspond to your local system.

## Simple Installation Method

This command will clean, build, and start the containers on your system. The only dependencies are Make and Docker.

```
make
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.1](https://github.com/michaelpeterswa/kochava_postback/compare/v1.1.0...v1.1.1) (2021-05-30)

## 1.1.0 (2021-05-29)


### Features

* added delivery/ingest folders and generated main.go + go-redis connection ([aa96e8c](https://github.com/michaelpeterswa/kochava_postback/commit/aa96e8c02fc26f29abd79f634bd6db215e26501f))
* added dockerfiles for delivery and ingest and started documentation ([c3a84f8](https://github.com/michaelpeterswa/kochava_postback/commit/c3a84f8e25a6e15d3284998c1fc6313307ea6b34))
* added environment variables ([ef02d45](https://github.com/michaelpeterswa/kochava_postback/commit/ef02d45b4c854e3f3f87b71fac78cd43e0a7c66c))
* added environment variables ([e0dbcae](https://github.com/michaelpeterswa/kochava_postback/commit/e0dbcae764cc01fe4846eb2307b95e2a48d7c495))
* added infinite queue w/ 5s empty timeout ([f438aab](https://github.com/michaelpeterswa/kochava_postback/commit/f438aab2214519dcaa6daf2447fa516723a708ee))
* added redis connection to php file ([02f352b](https://github.com/michaelpeterswa/kochava_postback/commit/02f352b2390d4e027285b7df13b6123d8066b011))
* created dockerfile and clean/run scripts ([fd1c84a](https://github.com/michaelpeterswa/kochava_postback/commit/fd1c84ae11e8c8c82c73c47a8b37fa7725ba0851))
* fixed scripting, added makefile, added regexes for defaults ([36ce6d5](https://github.com/michaelpeterswa/kochava_postback/commit/36ce6d5775bab506b2b54e3b96b321b3000df1dc))
* mvp working with php and go for ingest/delivery ([69bd5b4](https://github.com/michaelpeterswa/kochava_postback/commit/69bd5b4f121c1136fc313248cc710d79c62d723f))
* started php work ([e2964ef](https://github.com/michaelpeterswa/kochava_postback/commit/e2964efc4820956e2b4cc4c57dc44f7e1ffd92f0))


### Bug Fixes

* match scripts to documentation ([a6e1e3c](https://github.com/michaelpeterswa/kochava_postback/commit/a6e1e3c6398f8102585fd7f9ff240f10cb9548a7))
* match scripts to documentation ([76045d7](https://github.com/michaelpeterswa/kochava_postback/commit/76045d786eba756fde7223c6fdbdac471f04eefe))
* moved scripts into their own folder as to not pollute root ([b0152ba](https://github.com/michaelpeterswa/kochava_postback/commit/b0152bae0df7a08771cfba2d2a0f68c92c35b75d))
* refactored and split up monolithic main.go ([c8332f9](https://github.com/michaelpeterswa/kochava_postback/commit/c8332f9e777348c77cf5306b74fa9a8a617782d3))


delivery/postback.go
// Michael Peters
// postback.go
// Postback struct and handling for the Kochava Postback Delivery Agent
// Last Modified: 05/29/21 16:30 PDT

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/go-redis/redis/v8"
)

// https://stackoverflow.com/questions/36322141/unmarshal-inconsistent-json
// https://github.com/Jeffail/gabs
// https://bencane.com/2020/12/08/maps-vs-structs-for-json/

// the parent "postback" object
type Postback struct {
	Endpoint Endpoint    `json:"endpoint,omitempty"`
	Data     []DataValue `json:"data,omitempty"`
}

// the "endpoint" object based upon the ingested json
type Endpoint struct {
	Method string `json:"method,omitempty"`
	Url    string `json:"url,omitempty"`
}

// the "data" object based upon the ingested json
type DataValue struct {
	Mascot   string `json:"mascot,omitempty"`
	Location string `json:"location,omitempty"`
}

// the main loop that pops new pieces of data out of the Redis queue
func dequeueLoop(redisConn *redis.Client, key string) {
	// start infinite loop
	for {
		// one-second timeout on empty
		val, err := redisConn.BRPop(ctx, 1000000000, key).Result()
		if err == redis.Nil {
			// log.Printf("'%v' is currently empty", key)
		} else if err != nil {
			panic(err)
		} else {
			// need to get index[1] when using BRPop vs normal RPop because
			// the val returned looks like "[postback_queue test]" and we want "test"
			obj := val[1]
			log.Printf("Key/Val RPopped: %v:%v", key, val[1])
			if isValidJSON(obj) && obj != "null" {
				// log.Printf("-- Valid JSON --")
				start := time.Now()
				prepareResponse(obj, start)
			}
		}
	}
}

// using the object from the previous function, this unmarshals the string into
// a Postback object as defined above
func prepareResponse(data string, start time.Time) {
	var pb Postback

	err := json.Unmarshal([]byte(data), &pb)
	if err != nil {
		log.Println("error:", err)
	}

	for _, data := range pb.Data {
		// send the important pieces of data to the utility "constructURL" to perform
		// regex matching and replacement with the provided data
		respUrl := constructURL(pb.Endpoint.Url, data.Mascot, data.Location)
		sendResponse(respUrl, start)
	}
}

// once the response has been cleaned and transformed into a response URL, send request
func sendResponse(respUrl string, start time.Time) {
	resp, err := http.Get(respUrl)
	if err != nil {
		log.Println(err)
	} else {
		t := time.Now()
		elapsed := t.Sub(start)
		log.Printf("(%v) - time: %v - %v", resp.StatusCode, elapsed, respUrl)
		fmt.Printf("(%v) - time: %v - %v\n", resp.StatusCode, elapsed, respUrl)
	}
}
